src/xml_node.cc

```diff
@@ -230,68 +230,137 @@ XmlNode::New(xmlNode* node)
+    // increase the ref count of each parent
+    this->ref();
+
     XmlDocument* doc = static_cast<XmlDocument*>(xml_obj->doc->_private);
     doc->ref();
-
-    // check that there's a parent node with a _private pointer
-    // also check that the parent node isn't the doc since we already Ref() the doc once
-    if (xml_obj->parent != NULL &&
-        xml_obj->parent->_private != NULL &&
-        (void*)xml_obj->doc != (void*)xml_obj->parent)
-    {
-        static_cast<XmlNode*>(xml_obj->parent->_private)->Ref();
-    }
 }
 
 XmlNode::~XmlNode() {
 
-    // check if `xml_obj` has been freed so we don't access bad memory
-    if (this->freed)
-    {
-
-        // unref the doc using the doc reference we saved in the `flagNode` callback
-        if (this->doc)
-        {
-            XmlDocument* doc = static_cast<XmlDocument*>(this->doc->_private);
-            doc->unref();
-        }
+    // if we already freed xml_obj,
+    // then unref() the doc with this->doc
 
-        // set doc to null for good measure?
-        // this->doc = NULL;
-
-        // return so we don't attempt to use `xml_obj`
+    if (xml_obj == NULL) {
+        XmlDocument* doc = static_cast<XmlDocument*>(this->doc);
+        doc->unref();
         return;
     }
 
-    xml_obj->_private = NULL;
-    // release the hold and allow the document to be freed
+    // this->unref() may cause `xml_obj` to be freed
+    // so we create a reference to the document now
     XmlDocument* doc = static_cast<XmlDocument*>(xml_obj->doc->_private);
+
+    // decrease the ref count of each parent
+    this->unref();
+
+    // doc->unref() may cause `xml_obj` to be freed
+    // so we unref the document _after_ we unref each parent
     doc->unref();
 
-    // We do not free the xmlNode here if it is linked to a document
-    // It will be freed when the doc is freed
-    if (xml_obj->parent == NULL)
-      xmlFreeNode(xml_obj);
-
-    // if there's a parent then Unref() it
-    else if (xml_obj->parent->_private != NULL &&
-            (void*)xml_obj->doc != (void*)xml_obj->parent)
-    {
-        XmlNode* parent = static_cast<XmlNode*>(xml_obj->parent->_private);
-
-        // make sure Unref() is necessary
-        if (parent->refs_ > 0)
-        {
-            parent->Unref();
-        }
+    // unref() may have already set _private to NULL
+    // so we check that `xml_obj` is still available
+    if (xml_obj != NULL)
+        xml_obj->_private = NULL;
+}
+
+// recursively ref each parent
+void XmlNode::ref(int count) {
+    // make sure count > 0
+    if (count < 1)
+        count = 1;
+
+    // loop through each parent and increase the ref count
+    xmlNode* node = xml_obj;
+    while (node->parent != NULL) {
+        node = node->parent;
+        if (node->type == XML_DOCUMENT_NODE || node->type == XML_HTML_DOCUMENT_NODE)
+            break;
+        node->refs += count;
+    }
+}
+
+// recursively unref each parent
+void XmlNode::unref(int count) {
```

Comment:
I'm a little worried about this approach. Rather than explicitly counting references, why not hold an object which tracks referencing, similar to a smart pointer. This would be much safer than explicitly counting